custom_metrics/almanac.js

```diff
@@ -7,37 +7,233 @@
+        var node = nodes[i];
+        var item = node.getAttribute('itemtype');
+
+        if (item) {
+          // microdata
+          link.href = node.getAttribute('itemtype');
+          schemaElements[link.hostname + link.pathname] = true;
+        } else if ((node.tagName = 'SCRIPT')) {
+          // json+ld
+          try {
+            var content = JSON.parse(node.textContent);
+            var contentLoop = [];
+            if (content instanceof Object || content instanceof Array) {
+              // nested lookup
+              nestedLookup(content, 0);
+            }
+          } catch (e) {}
+        }
+      }
+    }
+    return Object.keys(schemaElements);
+  })(),
+  // Looks at links and identifies internal, external or hashed
+  'seo-anchor-elements': (() => {
+    // metric 10.10, 10.11,
+    var nodes = document.getElementsByTagName('a');
+    var link = document.createElement('a');
+
+    var internal = 0; // metric 10.10
+    var external = 0; // metric 10.10
+    var hash = 0;
+    var navigateHash = 0; // metric 10.11
 
     if (nodes) {
-        for (var i = 0, len = nodes.length; i < len; i++) {
-            var node = nodes[i];
-            var attributes = Object.values(node.attributes);
-            var el = {};
-
-            for (var n = 0, len2 = attributes.length; n < len2; n++) {
-                var attribute = attributes[n];
-                el[attribute.name.toLowerCase()] = attribute.value;
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+
+        // our local parser trick
+        if (node.href) {
+          link.href = node.href;
+
+          if (document.location.hostname === link.hostname) {
+            internal++;
+            if (document.location.pathname === link.pathname && link.hash.length > 1) {
+              // check if hash matches an element in the DOM (scroll to)
+              try {
+                var element = document.querySelector(link.hash);
+                if (element) {
+                  hash++;
+                } else {
+                  navigateHash++;
+                }
+              } catch (e) {}
             }
+          } else if (document.location.hostname !== link.hostname) {
+            external++;
+          }
+        }
+      }
+    }
+
+    return { internal, external, hash, navigateHash };
+  })(),
+  // Extracts titles used and counts the words, to flag thin content pages
+  'seo-titles': (() => {
+    //metric 10.9
+    var nodes = document.querySelectorAll('h1, h2, h3, h4');
+    var titleWords = -1;
+    var titleElements = -1;
+
+    if (nodes) {
+      titleWords = 0;
+      titleElements = 0;
 
-            linkNodes.push(el);
+      // analyse each title node
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+        // remove extra whitespace
+        var nodeText = node.textContent.replace(/\s+/g, ' ').replace(/^\s+|\s+$/, '');
```

Comment:
Rather than the second `replace`, use [trim()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/Trim)
```suggestion
        var nodeText = node.textContent.trim().replace(/\s+/g, ' ');
```